src/data/fintype/order.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2021 Peter Nelson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Peter Nelson
+-/
+
+import order.lattice
+import data.fintype.basic
+
+/-!
+# Order structures on finite sets
+
+This file shows that a `nonempty` `lattice` on a `fintype` gives a `complete_lattice`, and that a
+`linear_order` on a nonempty `fintype` gives a `complete_linear_order`. The latter applies in
+particular to `fin (n+1)`. Getting to a `bounded_lattice` from a `lattice` is computable, but the
+subsequent definitions are not, since the definitions of `Sup` and `Inf` use `set.to_finset`, which
+implicitly requires a `decidable_pred` instance for every `s : set α`.
+
+An explicit instance is given for a `complete_lattice` on `fin (n+1)`, but the rest are given
+as `def`s, to avoid loops in instance searches.
+-/
+
+section main
+
+/-- The maximum element in a `fintype` with `semilattice_sup` -/
+def fintype.top (α : Type*) [inhabited α] [fintype α] [semilattice_sup α] : α :=
+  finset.fold (⊔) (arbitrary α) id finset.univ
+
+/-- The minimum element in a `fintype` with `semilattice_inf` -/
+def fintype.bot (α : Type*) [inhabited α] [fintype α] [semilattice_inf α] : α :=
+  finset.fold (⊓) (arbitrary α) id finset.univ
+
+lemma fintype.bot_le (α : Type*) [inhabited α] [fintype α] [semilattice_inf α] (a : α) :
+  fintype.bot α ≤ a :=
+((@finset.fold_op_rel_iff_and _ _ (⊓) _ _ _ _ _ _
+  (λ _ _ _, le_inf_iff) (fintype.bot α)).mp le_rfl).2 a (finset.mem_univ _)
+
+lemma fintype.exists_bot (α : Type*) [i : nonempty α] [fintype α] [semilattice_inf α] :
+  ∃ m, ∀ a : α, m ≤ a :=
+⟨@fintype.bot α ⟨i.some⟩ _ _, @fintype.bot_le α ⟨i.some⟩ _ _⟩
+
+lemma fintype.eq_bot_iff (α : Type*) [inhabited α] [fintype α] [semilattice_inf α] (a : α) :
+  a = fintype.bot α ↔ ∀ x, a ≤ x :=
+⟨by {rintro rfl, apply fintype.bot_le}, λ h, le_antisymm (h _) (fintype.bot_le _ _)⟩
+
+lemma fintype.le_top (α : Type*) [inhabited α] [fintype α] [semilattice_sup α] (a : α) :
+  a ≤ fintype.top α :=
+(((@finset.fold_op_rel_iff_and _ _ (⊔) _ _ _ (arbitrary α) _ (λ x y, y ≤ x)
+  (λ _ _ _, sup_le_iff) (fintype.top α))).mp le_rfl).2 a (finset.mem_univ a)
+
+lemma fintype.exists_top (α : Type*) [i : nonempty α] [fintype α] [semilattice_sup α] :
+  ∃ m, ∀ a : α, a ≤ m :=
+⟨@fintype.top α ⟨i.some⟩ _ _, @fintype.le_top α ⟨i.some⟩ _ _⟩
+
+lemma fintype.eq_top_iff (α : Type*) [inhabited α] [fintype α] [semilattice_sup α] (a : α) :
+  a = fintype.top α ↔ ∀ x, x ≤ a :=
+⟨by {rintro rfl, apply fintype.le_top}, λ h, le_antisymm (fintype.le_top _ _) (h _)⟩
+
+/-- A nonempty finite lattice is bounded  -/
+def fintype.bounded_lattice (α : Type*) [inhabited α] [fintype α] [lattice α] :
+  bounded_lattice α :=
+{ bot := fintype.bot α,
+  bot_le := fintype.bot_le α,
+  top := fintype.top α,
+  le_top := fintype.le_top α,
+  .. (infer_instance : lattice α)}
+
```

Comment:
Right now there are two ways to refer to the top element of a fintype with lattice structure: `fintype.top` or `⊤`: I think there should be a simp lemma saying `fintype.top = ⊤` - that way things like `eq_top_iff` follow from general results on bounded lattices.

Reply:
Thanks. This made me realize that defining `top`/`bot` outside structures didn't make sense in the first place. Latest push just upgrades `semilattice_sup` to `semilattice_sup_top` directly to gain access to `top` notation.